enhancements/network/dual-stack.md

```diff
@@ -0,0 +1,236 @@
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+Customers need IPv6 support. Customers still need IPv4 support too
+though. Dual-stack FTW!
+
+At the moment we expect dual-stack to be more popular with customers
+than single-stack IPv6, even for customers who mostly only want IPv6,
+because in many cases you end up needing 99% IPv6 and 1% IPv4, and
+single-stack IPv6 Kubernetes does not allow that. (You end up needing
+hacks outside of Kubernetes's view, like Multus-based interfaces.)
+
+## Motivation
+
+Customers have requested this functionality.
+
+### Goals
+
+- All known upstream Kubernetes dual-stack blockers will be resolved.
+  (eg, the Service `ipFamily` problem that blocked us from enabling
+  the `IPv6DualStack` feature gate in 4.4).
+
+- OVN-Kubernetes will support dual-stack networking according to the
+  model upstream Kubernetes finalizes
+
+- We enable the `IPv6DualStack` feature gate by default
+
+- Other OCP components that need to be updated to be dual-stack aware
+  will be updated. (It is not clear how much actual work is needed
+  here. We are currently researching.)
+
+- We will have CI/cluster-bot support for dual-stack clusters on
+  platforms that we support for customers (ie, bare-metal).
+
+- Assuming that the necessary CloudProvider work is completed
+  upstream, we will have CI/cluster-bot support for dual-stack
+  clusters on at least one cloud platform. (If the bare-metal solution
+  is not sufficiently useful/available, we might do some work to help
+  complete the upstream CloudProvider work.)
+
+### Non-Goals
+
+- Dual-stack support for openshift-sdn or Kuryr
+
+- Officially supporting dual-stack for third-party CNI plugins
+
+## Proposal
+
+### Implementation Details/Notes/Constraints
+
+#### Early CI
+
+To avoid the problems we had with single-stack IPv6, we will get CI
+set up early, to ensure that regressions do not occur in components we
+are not directly paying attention to. Note that we do not expect to do
+any work on a forked branch of OCP as we did for single-stack IPv6.
+
+Being able to bring up a dual-stack cluster at all requires enabling
+the `IPv6DualStack` feature gate in `kubelet`, `kube-apiserver`, and
+`kube-controller-manager`. We tried to do this for all clusters in 4.4
+but [ran into bugs] and had to revert it. The [upstream bugs] are not
+yet fixed.
+
+To work around this for early CI, we will patch
+`machine-config-operator`, `cluster-kube-apiserver-operator`, and
+`cluster-kube-controller-manager-operator` to add the feature gate to
+their operands if and only if the `install-config.yaml` defines a
+dual-stack cluster. (This won't allow migrating from single-stack to
+dual-stack but testing that can wait until after the upstream problems
+are fixed and we can enable the feature gate everywhere.)
```

Comment:
mco/master people, does this sound like a reasonable plan?
eg @cgwalters @runcom @deads2k @sttts @soltysh

Reply:
Sounds reasonable.

Reply:
Went to try to implement this and discovered that we can just add a `FeatureGate.config.openshift.io` object to the install manifests to override it without needing any code changes, and that will probably be simpler.